mchf-eclipse/drivers/audio/audio_driver.c

```diff
@@ -4777,28 +4807,35 @@ static void AudioDriver_TxProcessor(AudioSample_t * const srcCodec, IqSample_t *
         }
     }
 
-#ifdef UI_BRD_OVI40
-    // we code the sidetone to the audio codec, since we have one for audio and one for iq
-    // TODO: Do we need the check? OVI40 always comes with 2 codecs
-    if ((void*)audioDst != (void*)dst)
-    {
-    	if (dmod_mode == DEMOD_CW)
-    	{
-    		for (uint16_t idx = 0; idx < blockSize; idx++)
-    		{
-    			audioDst[idx].r = audioDst[idx].l = dst[idx].l;
-    		}
-    	} else if (is_demod_rtty())
-    	{
```

Comment:
For RTTY and BPSK but not for CW, but it's a good idea, thanks I will think about it.
ts.tx_power_factor shouldn't be used for audio output.

Reply:
Maybe change CW to use a_buffer for generating sine, so all modes will have one source of signal for audio?

Reply:
Are you going to change AudioDriver_TxIqProcessingFinal() that it will receive source as a parameter? 

Another notice, we have TX_Audio_via_USB in "Debug settings" -> Source, Filtered and Final IQ, so, in many modes the source buffer modified during processing. Maybe we need dedicate fixed buffer for the source,  What's a difference between Filtered and final IQ?

Reply:
>
> 
> Maybe change CW to use a_buffer for generating sine, so all modes will have one source of signal for audio?

I was thinking of this.



> 
> 
> Are you going to change AudioDriver_TxIqProcessingFinal() that it will receive source as a parameter?
>

Yes, finally.
 
> Another notice, we have TX_Audio_via_USB in "Debug settings" -> Source, Filtered and Final IQ, so, in many modes the source buffer modified during processing. Maybe we need dedicate fixed buffer for the source, What's a difference between Filtered and final IQ?
It is not filtered IQ, it is filter audio input (!) after filtering and compression. It is in fact only FM where the a_buffer is being "destroyed" and if we really want, we could use a_buffer[1] for that, so that a_buffer[0] remains intact.

Reply:
> > Are you going to change AudioDriver_TxIqProcessingFinal() that it will receive source as a parameter?
> 
> Yes, finally.
> 
Ok, in this case, I would like to change buffer for generated CW wave to a_buffer and add HACK into 
AudioDriver_TxIqProcessingFinal() for this and later me or you change when parameter would be added. 

That gives to put filling audio buffer in one place is it was before.

Reply:
I don't see why you have to  change AudioDriver_TxIqProcessingFinal() , you can simply copy the i_buffer (and scale a_buffer) before the AudioDriver_TxIqProcessingFinal is being run. In case of signal_active == false you have to fill the a_buffer probably with silence.

Reply:
One or another, doesn't matter.. Ok.